src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/FederationAccessManagerWithCache.java

```diff
@@ -0,0 +1,111 @@
+import se.liu.ida.hefquin.engine.federation.access.StringResponse;
+import se.liu.ida.hefquin.engine.federation.access.TPFRequest;
+import se.liu.ida.hefquin.engine.federation.access.TPFResponse;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+
+public class FederationAccessManagerWithCache implements FederationAccessManager 
+{
+	private FederationAccessManager fedAccMan;
+	private Map<TriplePattern, CompletableFuture<CardinalityResponse>> cacheMap;
+	//private Map<TriplePattern, Integer> cacheMap;
+
+	public FederationAccessManagerWithCache() {
+		System.out.println("Inside normal constructor");
+	}
+	
+	public FederationAccessManagerWithCache(FederationAccessManager fedAccMan) {
+		this.fedAccMan = fedAccMan;
+		//this.cacheMap = new HashMap<TriplePattern, Integer>();
+		this.cacheMap = new HashMap<TriplePattern, CompletableFuture<CardinalityResponse>>();
+		System.out.println("Inside constructor creating fedAccMan and cachemap");
+	}
+	
+	public void addResultToCache(TriplePattern tp, CompletableFuture<CardinalityResponse> response) {
+		cacheMap.put(tp, response);
+	}
+	
+	/*public void addResultToCache(TriplePattern tp, Integer response) {
+		cacheMap.put(tp, response);
+	}*/
+
+	@Override
+	public CompletableFuture<SolMapsResponse> issueRequest(SPARQLRequest req, SPARQLEndpoint fm)
+			throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<TPFResponse> issueRequest(TPFRequest req, TPFServer fm) throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<TPFResponse> issueRequest(TPFRequest req, BRTPFServer fm)
+			throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<TPFResponse> issueRequest(BRTPFRequest req, BRTPFServer fm)
+			throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<StringResponse> issueRequest(Neo4jRequest req, Neo4jServer fm)
+			throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<CardinalityResponse> issueCardinalityRequest(SPARQLRequest req, SPARQLEndpoint fm)
+			throws FederationAccessException {
+		return fedAccMan.issueCardinalityRequest(req, fm);
+	}
+
+	// TODO Add caching for this request
+	@Override
+	public CompletableFuture<CardinalityResponse> issueCardinalityRequest(TPFRequest req, TPFServer fm)
+			throws FederationAccessException {
+		System.out.println(cacheMap.values());
+		//Integer cacheResponse = cacheMap.get(req.getQueryPattern());
+		CompletableFuture<CardinalityResponse> cacheResponse = cacheMap.get(req.getQueryPattern());
+		if (cacheResponse == null) {
+			System.out.println("Did not find in cache - running request");
+			return fedAccMan.issueCardinalityRequest(req, fm).thenApply((response) -> addResultToCache(req.getQueryPattern(), response));
```

Comment:
It is actually easier than that. Simply take the `CompletableFuture<CardinalityResponse>` obtained from `issueCardinalityRequest` and add it directly into the cache, and then return it.
```java
final CompletableFuture<CardinalityResponse> newFutureResponse = fedAccMan.issueCardinalityRequest(req, fm);
addResultToCache(req.getQueryPattern(), newFutureResponse);
return newFutureResponse;
```
I understand that the futures API of Java (i.e., `CompletableFuture` etc.) is a bit tricky to get your head around and it is not typically a topic in the university courses. I can explain a bit more about it in our next meeting.